compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,457 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "astutil.h"
+#include "baseAST.h"
+#include "stlUtil.h"
+#include "expr.h"
+#include "stmt.h"
+#include "symbol.h"
+#include "type.h"
+#include "LoopStmt.h"
+#include "CForLoop.h"
+#include "WhileStmt.h"
+#include "exprAnalysis.h"
+
+//helper datastructures/types
+
+typedef std::map<SymExpr*, std::pair<Expr*,Type*> > UseDefCastMap;
+
+//prototypes
+bool primMoveGeneratesCommCall(CallExpr* ce);
+inline bool unsafeExprInBetween(Expr* e1, Expr* e2);
+inline bool requiresCast(Type* t);
+inline bool isIntegerPromotionPrimitive(PrimitiveTag tag);
+bool isDenormalizable(Symbol* sym,
+    Map<Symbol*,Vec<SymExpr*>*>& defMap,
+    Map<Symbol*,Vec<SymExpr*>*>& useMap, SymExpr** useOut, Expr** defOut,
+    Type** castTo);
+void findCandidatesInFunc(FnSymbol *fn, UseDefCastMap& candidates);
+void findCandidatesInFunc(FnSymbol* fn, Vec<Symbol*> symVec,
+    UseDefCastMap& udcMap);
+void denormalize(void);
+void denormalize(Expr* def, SymExpr* use, Type* castTo);
+void denormalizeOrDeferCandidates(UseDefCastMap& candidates,
+    Vec<Symbol*>& deferredSyms);
+
+/*
+ * This function tries to remove temporary variables in function `fn`
+ *
+ * A local variable is removed if:
+ *
+ * - It is def'd and use'd once
+ *
+ * - Its def is a PRIM_MOVE or PRIM_ASSIGN, with no possible communication
+ * - RHS and LHS are of same type and non-extern
+ *
+ * - Its use a suitable primitive
+ * - Its use is not repeated(condition/increment statement of a loop)
+ *
+ * Denormalization uses helpers in util/exprAnalysis.cpp to decide if it's safe
+ * to move function calls and primitives.
+ */
+void denormalize(void) {
+
+  UseDefCastMap candidates;
+  Vec<Symbol*> deferredSyms;
+
+  if(fDenormalize) {
+    forv_Vec(FnSymbol, fn, gFnSymbols) {
+      bool isFirstRound = true;
+      do {
+        candidates.clear();
+        deferredSyms.clear();
+
+        // if we are analyzing locals in a function for the first time, we look
+        // at all the symbols in the function. otherwise we only look at the
+        // ones deferred in the previous passes on the same function
+        if(isFirstRound) {
+          findCandidatesInFunc(fn, candidates);
+        }
+        else {
+          findCandidatesInFunc(fn, deferredSyms, candidates);
+        }
+        denormalizeOrDeferCandidates(candidates, deferredSyms);
+
+        isFirstRound = false;
+      } while(deferredSyms.count() > 0);
```

Comment:
I would have expected this loop to terminate only after denormalization took no action (that is, the number of denormalized symbols was 0).